src/Ng/Program.cs

```diff
@@ -56,62 +58,107 @@ public static async Task MainAsync(string[] args)
 
-                // Create a logger that is scoped to this class (only)
-                _logger = loggerFactory.CreateLogger<Program>();
+                // This tells Application Insights that, even though a heartbeat is reported, 
+                // the state of the application is unhealthy when the exitcode is different from zero.
+                // The heartbeat metadata is enriched with the job loop exit code.
+                applicationInsightsConfiguration.DiagnosticsTelemetryModule?.AddOrSetHeartbeatProperty(
+                    HeartbeatProperty_JobLoopExitCode,
+                    exitCode.ToString(),
+                    isHealthy: exitCode == 0);
 
                 var cancellationTokenSource = new CancellationTokenSource();
-
-                // Create an ITelemetryService
-                var telemetryService = new TelemetryService(new TelemetryClient());
-
-                // Allow jobs to set global custom dimensions
-                TelemetryConfiguration.Active.TelemetryInitializers.Add(new JobPropertiesTelemetryInitializer(telemetryService));
-
-                job = NgJobFactory.GetJob(jobName, telemetryService, loggerFactory);
                 await job.RunAsync(arguments, cancellationTokenSource.Token);
+                exitCode = 0;
             }
             catch (ArgumentException ae)
             {
+                exitCode = 1;
                 _logger?.LogError("A required argument was not found or was malformed/invalid: {Exception}", ae);
 
                 Console.WriteLine(job != null ? job.GetUsage() : NgJob.GetUsageBase());
             }
             catch (KeyNotFoundException knfe)
             {
+                exitCode = 1;
                 _logger?.LogError("An expected key was not found. One possible cause of this is required argument has not been provided: {Exception}", knfe);
 
                 Console.WriteLine(job != null ? job.GetUsage() : NgJob.GetUsageBase());
             }
             catch (Exception e)
             {
+                exitCode = 1;
                 _logger?.LogCritical("A critical exception occured in ng.exe! {Exception}", e);
             }
 
+            applicationInsightsConfiguration.DiagnosticsTelemetryModule?.SetHeartbeatProperty(
+                HeartbeatProperty_JobLoopExitCode,
+                exitCode.ToString(),
+                isHealthy: exitCode == 0);
+
             Trace.Close();
-            TelemetryConfiguration.Active.TelemetryChannel.Flush();
+            applicationInsightsConfiguration?.TelemetryConfiguration.TelemetryChannel.Flush();
+        }
+
+        private static ILoggerFactory ConfigureLoggerFactory(ApplicationInsightsConfiguration applicationInsightsConfiguration)
+        {
+            var loggerConfiguration = LoggingSetup.CreateDefaultLoggerConfiguration(withConsoleLogger: true);
+            loggerConfiguration.WriteTo.File("Log.txt", retainedFileCountLimit: 3, fileSizeLimitBytes: 1000000, rollOnFileSizeLimit: true);
```

Comment:
Do we really need file logging? The volume is so high that it's rather hard to use in many cases.

Reply:
This is not a new addition to the code. It's what we currently do. See https://github.com/NuGet/NuGet.Services.Metadata/blob/master/src/Ng/Program.cs#L78.

If we don't need it nor use it, we should indeed remove this. So do we ever us this?

Reply:
Oops forgot it was preexisting. Out of scope.